ethminer: init at 0.18.0-rc.0

nand0p opened the pull request:
- updates jsoncpp
- adds ethash and cli11
- tested on NixOS

symphorien commented:
Have you tried without hunter ? https://github.com/ethereum-mining/ethminer/blob/master/docs/BUILD.md#disable-hunter

The dependencies they list (cmake and others) seem reasonable.

symphorien commented on pkgs/tools/misc/ethminer/default.nix:
```diff
@@ -0,0 +1,26 @@
+    description = "Ethereum miner with OpenCL, CUDA and stratum support";
+    homepage = https://github.com/ethereum-mining/ethminer;
+    platforms = [ "x86_64-linux" ];
+    maintainers = with maintainers; [ nand0p ];
```
please indicate the license

nand0p replied:
done

nand0p commented:
> Have you tried without hunter ? https://github.com/ethereum-mining/ethminer/blob/master/docs/BUILD.md#disable-hunter

i was not successful in decoupling hunter. it appears `DHUNTER_ENABLED=OFF` did not work for me. perhaps someone more skilled in nix could fixup this code iteratively.

symphorien commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,96 @@
+
+      pool = mkOption {
+        type = types.string;
+        default = "mining pool address";
```
Is this a valid default value ?

nand0p replied:
updated to valid defaults

symphorien commented:
You mean:
```
cmakeFlags = [ "-DHUNTER_ENABLED=OFF" ];
``` ?

nand0p commented:
> You mean:
> 
> ```
> cmakeFlags = [ "-DHUNTER_ENABLED=OFF" ];
> ```
> ?

apparently, that doesnt seem to work.  https://github.com/ethereum-mining/ethminer/issues/500

```
CMake Error at CMakeLists.txt:64 (find_package):
  Could not find a package configuration file provided by "jsoncpp" with any
  of the following names:

    jsoncppConfig.cmake
    jsoncpp-config.cmake

  Add the installation prefix of "jsoncpp" to CMAKE_PREFIX_PATH or set
  "jsoncpp_DIR" to a directory containing one of the above files.  If
  "jsoncpp" provides a separate development package or SDK, be sure it has
  been installed.
```

symphorien commented:
here is an expression which contains `jsoncpp-config.cmake`:
```nix
jsoncpp.overrideAttrs(old: {
  cmakeFlags= (old.cmakeFlags or []) ++ ["-DJSONCPP_WITH_CMAKE_PACKAGE=ON"];
})
```
This change seems benign enough to me that you can change it globally in  pkgs/development/libraries/jsoncpp/default.nix instead of overriding it only for ethminer.

nand0p commented:
```
$ nix-build -A ethminer     
these derivations will be built:
  /nix/store/jy4m3y3bvwj1v0vm7ciy3ydrbj4svi2j-ethminer-0.17.0.drv
building '/nix/store/jy4m3y3bvwj1v0vm7ciy3ydrbj4svi2j-ethminer-0.17.0.drv'...
unpacking sources
unpacking source archive /nix/store/myajjwbj782daxz6h69j9c6m4dq9b1m8-source
source root is source
patching sources
configuring
/build/source/build /build/source
-- [cable ] Cable 0.2.14 initialized
-- The C compiler identification is GNU 7.4.0
-- The CXX compiler identification is GNU 7.4.0
-- Check for working C compiler: /nix/store/bzxfiz7l38g7b4d8rdlf4wc43fklah5h-gcc-wrapper-7.4.0/bin/gcc
-- Check for working C compiler: /nix/store/bzxfiz7l38g7b4d8rdlf4wc43fklah5h-gcc-wrapper-7.4.0/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /nix/store/bzxfiz7l38g7b4d8rdlf4wc43fklah5h-gcc-wrapper-7.4.0/bin/g++
-- Check for working CXX compiler: /nix/store/bzxfiz7l38g7b4d8rdlf4wc43fklah5h-gcc-wrapper-7.4.0/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- [cable ] Build type: Release
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Boost version: 1.67.0
-- Found the following Boost libraries:
--   system
--   filesystem
--   thread
--   chrono
--   date_time
--   atomic
----------------------------------------------------------------------------
-- CMake 3.12.1
-- Build Release / Linux
----------------------------------------------------------------- components
-- ETHASHCL         Build OpenCL components                      ON
-- ETHASHCUDA       Build CUDA components                        ON
-- ETHASHCPU        Build CPU components (only for development)  OFF
-- ETHDBUS          Build D-Bus components                       ON
-- APICORE          Build API Server components                  ON
-- BINKERN          Install AMD binary kernels                   ON
-- DEVBUILD         Build with dev logging                       OFF
----------------------------------------------------------------------------

-- Performing Test ffunction-sections
-- Performing Test ffunction-sections - Success
-- Performing Test fdata-sections
-- Performing Test fdata-sections - Success
-- Performing Test Wl,--gc-sections
-- Performing Test Wl,--gc-sections - Success
-- Found Git: /nix/store/qyp22r9zfhpiykkj57rj6jcd6iwgdccy-git-2.19.2/bin/git (found version "2.19.2")
-- Found CUDA: /nix/store/qqpzxqw5nxxqyxlbjb9rw4sx8m2nfy8m-cudatoolkit-9.2.148.1 (found version "9.2")
-- Found OpenSSL: /nix/store/zxi65ww7bws5lqlxgljq0p2j6pd4waiv-openssl-1.0.2q/lib/libcrypto.so (found version "1.0.2q")
-- Looking for CL_VERSION_2_2
-- Looking for CL_VERSION_2_2 - found
-- Found OpenCL: /nix/store/cp6nlngvrhs5fbbkjh4c4dpafxzrxr8d-ocl-icd-2.2.10/lib/libOpenCL.so (found version "2.2")
-- Found PkgConfig: /nix/store/zq9pg3xg7d1v48ac10icw2li8qyjwqiz-pkg-config-0.29.2/bin/pkg-config (found version "0.29.2")
-- Checking for module 'dbus-1'
--   Found dbus-1, version 1.12.12
-- Configuring done
CMake Error in libethcore/CMakeLists.txt:
  Imported target "ethash::ethash" includes non-existent path

    "/nix/store/4bhcm2l7385z8rpbl1lkjpq61r6pppip-ethash-0.4.2/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in libethcore/CMakeLists.txt:
  Imported target "ethash::ethash" includes non-existent path

    "/nix/store/4bhcm2l7385z8rpbl1lkjpq61r6pppip-ethash-0.4.2/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in libpoolprotocols/CMakeLists.txt:
  Imported target "ethash::ethash" includes non-existent path

    "/nix/store/4bhcm2l7385z8rpbl1lkjpq61r6pppip-ethash-0.4.2/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in libpoolprotocols/CMakeLists.txt:
  Imported target "ethash::ethash" includes non-existent path

    "/nix/store/4bhcm2l7385z8rpbl1lkjpq61r6pppip-ethash-0.4.2/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in ethminer/CMakeLists.txt:
  Imported target "CLI11::CLI11" includes non-existent path

    "/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in ethminer/CMakeLists.txt:
  Imported target "CLI11::CLI11" includes non-existent path

    "/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



-- Generating done
-- Build files have been written to: /build/source/build
builder for '/nix/store/jy4m3y3bvwj1v0vm7ciy3ydrbj4svi2j-ethminer-0.17.0.drv' failed with exit code 1
error: build of '/nix/store/jy4m3y3bvwj1v0vm7ciy3ydrbj4svi2j-ethminer-0.17.0.drv' failed
```
ethash and cli11 are apparently not exporting the correct include path. any ideas here?

symphorien commented on pkgs/development/libraries/ethash/default.nix:
```diff
@@ -0,0 +1,54 @@
+  buildPhase = ''
+    cmake . $cmakeFlags
+    cmake --build .
+  '';
```
I would expect the build to work the same with the default build phase.

symphorien commented on pkgs/development/libraries/ethash/default.nix:
```diff
@@ -0,0 +1,54 @@
+    };
+
+  buildInputs = [
+    cmake
```
cmake should be a nativeBuildInput

symphorien commented on pkgs/development/libraries/ethash/default.nix:
```diff
@@ -0,0 +1,54 @@
+    ocl-icd
+    mesa
+    gbenchmark
+    gtest
```
test dependencies should go in checkInputs

symphorien commented on pkgs/development/tools/misc/cli11/default.nix:
```diff
@@ -0,0 +1,34 @@
+
+  buildPhase = ''
+    cmake . $cmakeFlags
+  '';
```
try to keep the default buildPhase

symphorien commented on pkgs/tools/misc/ethminer/default.nix:
```diff
@@ -0,0 +1,83 @@
+
+  configurePhase = ''
+    mkdir -p build/install
+    pushd build
```
use preConfigure for this, and then leave the  default configurePhase

symphorien commented:
I think the problem is that you override the default configure, build and or installPhase.
As soon as cmake is in nativeBuildInputs (or buildInputs, but don't do that) the default phases will
use cmake automagically for configuring, building and installing. This will result in the correct hierarchy in $out:
with this diff
```diff
diff --git a/pkgs/development/tools/misc/cli11/default.nix b/pkgs/development/tools/misc/cli11/default.nix
index 47c5b7e2a9d..258e5af0ee7 100644
--- a/pkgs/development/tools/misc/cli11/default.nix
+++ b/pkgs/development/tools/misc/cli11/default.nix
@@ -13,22 +13,11 @@ stdenv.mkDerivation rec {
     #fetchSubmodules = true;
   };
 
-  buildInputs = [
+  nativeBuildInputs = [
     cmake
   ];
 
   cmakeFlags = [
     "-DCLI11_TESTING=OFF"
   ];
-
-  buildPhase = ''
-    cmake . $cmakeFlags
-  '';
-
-  installPhase = ''
-    mkdir -pv $out
-    cp -rv * $out
-    cp -v CMakeFiles/Export/lib/cmake/CLI11/CLI11Config.cmake $out/cli11-config.cmake
-  '';
-
 }
```
I get the very clean hierarchy
```
result
├── include
│   └── CLI
│       ├── App.hpp
│       ├── CLI.hpp
│       ├── ConfigFwd.hpp
│       ├── Config.hpp
│       ├── Error.hpp
│       ├── FormatterFwd.hpp
│       ├── Formatter.hpp
│       ├── Macros.hpp
│       ├── Optional.hpp
│       ├── Option.hpp
│       ├── Split.hpp
│       ├── StringTools.hpp
│       ├── Timer.hpp
│       ├── TypeTools.hpp
│       ├── Validators.hpp
│       └── Version.hpp
└── lib
    └── cmake
        └── CLI11
            ├── CLI11Config.cmake
            └── CLI11ConfigVersion.cmake
```
for cli11 instead of 
```
result
├── cli11-config.cmake
├── CLI11ConfigVersion.cmake
├── CLI11Targets.cmake
├── CMakeCache.txt
├── CMakeFiles
│   ├── 3.12.1
│   │   ├── CMakeCXXCompiler.cmake
│   │   ├── CMakeDetermineCompilerABI_CXX.bin
│   │   ├── CMakeSystem.cmake
│   │   └── CompilerIdCXX
│   │       ├── a.out
│   │       ├── CMakeCXXCompilerId.cpp
│   │       └── tmp
│   ├── cmake.check_cache
│   ├── CMakeDirectoryInformation.cmake
│   ├── CMakeOutput.log
│   ├── CMakeTmp
│   ├── Export
│   │   └── lib
│   │       └── cmake
│   │           └── CLI11
│   │               └── CLI11Config.cmake
│   ├── feature_tests.bin
│   ├── feature_tests.cxx
│   ├── Makefile2
│   ├── Makefile.cmake
│   ├── progress.marks
│   └── TargetDirectories.txt
├── cmake_install.cmake
├── CPackConfig.cmake
├── CPackSourceConfig.cmake
├── examples
│   ├── CMakeFiles
│   │   ├── CMakeDirectoryInformation.cmake
│   │   ├── enum.dir
│   │   │   ├── build.make
│   │   │   ├── cmake_clean.cmake
│   │   │   ├── DependInfo.cmake
│   │   │   ├── depend.make
│   │   │   ├── flags.make
│   │   │   ├── link.txt
│   │   │   └── progress.make
│   │   ├── formatter.dir
│   │   │   ├── build.make
│   │   │   ├── cmake_clean.cmake
│   │   │   ├── DependInfo.cmake
│   │   │   ├── depend.make
│   │   │   ├── flags.make
│   │   │   ├── link.txt
│   │   │   └── progress.make
│   │   ├── groups.dir
│   │   │   ├── build.make
│   │   │   ├── cmake_clean.cmake
│   │   │   ├── DependInfo.cmake
│   │   │   ├── depend.make
│   │   │   ├── flags.make
│   │   │   ├── link.txt
│   │   │   └── progress.make
│   │   ├── inter_argument_order.dir
│   │   │   ├── build.make
│   │   │   ├── cmake_clean.cmake
│   │   │   ├── DependInfo.cmake
│   │   │   ├── depend.make
│   │   │   ├── flags.make
│   │   │   ├── link.txt
│   │   │   └── progress.make
│   │   ├── modhelp.dir
│   │   │   ├── build.make
│   │   │   ├── cmake_clean.cmake
│   │   │   ├── DependInfo.cmake
│   │   │   ├── depend.make
│   │   │   ├── flags.make
│   │   │   ├── link.txt
│   │   │   └── progress.make
│   │   ├── nested.dir
│   │   │   ├── build.make
│   │   │   ├── cmake_clean.cmake
│   │   │   ├── DependInfo.cmake
│   │   │   ├── depend.make
│   │   │   ├── flags.make
│   │   │   ├── link.txt
│   │   │   └── progress.make
│   │   ├── prefix_command.dir
│   │   │   ├── build.make
│   │   │   ├── cmake_clean.cmake
│   │   │   ├── DependInfo.cmake
│   │   │   ├── depend.make
│   │   │   ├── flags.make
│   │   │   ├── link.txt
│   │   │   └── progress.make
│   │   ├── progress.marks
│   │   ├── simple.dir
│   │   │   ├── build.make
│   │   │   ├── cmake_clean.cmake
│   │   │   ├── DependInfo.cmake
│   │   │   ├── depend.make
│   │   │   ├── flags.make
│   │   │   ├── link.txt
│   │   │   └── progress.make
│   │   ├── subcom_help.dir
│   │   │   ├── build.make
│   │   │   ├── cmake_clean.cmake
│   │   │   ├── DependInfo.cmake
│   │   │   ├── depend.make
│   │   │   ├── flags.make
│   │   │   ├── link.txt
│   │   │   └── progress.make
│   │   ├── subcommands.dir
│   │   │   ├── build.make
│   │   │   ├── cmake_clean.cmake
│   │   │   ├── DependInfo.cmake
│   │   │   ├── depend.make
│   │   │   ├── flags.make
│   │   │   ├── link.txt
│   │   │   └── progress.make
│   │   └── validators.dir
│   │       ├── build.make
│   │       ├── cmake_clean.cmake
│   │       ├── DependInfo.cmake
│   │       ├── depend.make
│   │       ├── flags.make
│   │       ├── link.txt
│   │       └── progress.make
│   ├── cmake_install.cmake
│   ├── Makefile
│   └── subcom_in_files
│       ├── CMakeFiles
│       │   ├── CMakeDirectoryInformation.cmake
│       │   ├── progress.marks
│       │   └── subcommand_main.dir
│       │       ├── build.make
│       │       ├── cmake_clean.cmake
│       │       ├── DependInfo.cmake
│       │       ├── depend.make
│       │       ├── flags.make
│       │       ├── link.txt
│       │       └── progress.make
│       ├── cmake_install.cmake
│       └── Makefile
└── Makefile
```
Note how the automatic phases have installed the missing /include directory.

I expect that if you remove all your custom buildPhases, configurePhases and installPhases, everything should work well.
In case the default installPhase misses a file (for example) check if cmake has flags to install it, or use `postInstall` to install the file manually, instead of overriding the whole installPhase.
I hope this is helpful.

nand0p commented:
ethminer systemd service:
```
● ethminer.service - ethminer ethereum mining service
   Loaded: loaded (/nix/store/rdsc442ydk9xa67g2gll8iqi5v5l5my5-unit-ethminer.service/ethminer.service; enabled; vendor preset: enabled)
   Active: active (running) since Sun 2019-02-17 11:21:44 EST; 6min ago
 Main PID: 30801 (c5p7n314ffm0f8f)
    Tasks: 11 (limit: 4915)
   Memory: 73.8M
   CGroup: /system.slice/ethminer.service
           ├─30801 /nix/store/i82x3x0yiijkgyqkzh8ni87gspas0f48-bash-4.4-p23/bin/bash -e /nix/store/c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start
           └─30803 /nix/store/jbflf9xz40h5mia4sp8drsnb89by4r68-ethminer-0.18.0-alpha.3/bin/ethminer --farm-recheck 1000 --report-hashrate --cuda --api-port -3333 --pool stratum1+tcp://[REDACTED]

Feb 17 11:27:34 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  m 11:27:34 .ethminer-wrapp 0:05 A2 53.55 Mh { cu0 26.70 | cu1 26.85 }
```

ethminer journal:
```
Feb 17 11:21:44 koko systemd[1]: Starting ethminer ethereum mining service...
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]: ethminer 0.18.0-alpha.3
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]: Build: linux/release/gnu
Feb 17 11:21:44 koko s5gg2q5xalifqmb9zqjcd36h3vb7cf1x-unit-script-ethminer-post-start[30802]: Power limit for GPU 00000000:01:00.0 was set to 112.00 W from 110.00 W.
Feb 17 11:21:44 koko s5gg2q5xalifqmb9zqjcd36h3vb7cf1x-unit-script-ethminer-post-start[30802]: All done.
Feb 17 11:21:44 koko systemd[1]: Started ethminer ethereum mining service.
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  i 11:21:44 .ethminer-wrapp Configured pool eth-us-east1.nanopool.org:9999
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  i 11:21:44 .ethminer-wrapp Api server listening on port 3333.
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  i 11:21:44 .ethminer-wrapp Selected pool eth-us-east1.nanopool.org:9999
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  i 11:21:44 .ethminer-wrapp Established connection to eth-us-east1.nanopool.org  [192.99.69.170:9999]
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  i 11:21:44 .ethminer-wrapp Spinning up miners...
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]: cu 11:21:44 cuda-0   Using Pci Id : 01:00.0 GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  i 11:21:44 .ethminer-wrapp Stratum mode : ETHPROXY Compatible
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  i 11:21:44 .ethminer-wrapp Logged in!
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]: cu 11:21:44 cuda-1   Using Pci Id : 02:00.0 GeForce GTX 1070 (Compute 6.1) Memory : 7.93 GB
Feb 17 11:21:44 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  i 11:21:44 .ethminer-wrapp Epoch : 241 Difficulty : 10.00 Gh
Feb 17 11:21:46 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]: cu 11:21:46 cuda-0   Generating DAG + Light : 2.93 GB
Feb 17 11:21:46 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]: cu 11:21:46 cuda-1   Generating DAG + Light : 2.93 GB
Feb 17 11:21:48 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  i 11:21:48 .ethminer-wrapp Job: f79ef197… eth-us-east1.nanopool.org  [192.99.69.170:9999]
Feb 17 11:21:55 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]: cu 11:21:55 cuda-1   Generated DAG + Light in 9,180 ms. 5.00 GB left.
Feb 17 11:21:56 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]: cu 11:21:56 cuda-0   Generated DAG + Light in 9,929 ms. 5.00 GB left.
Feb 17 11:22:04 koko c5p7n314ffm0f8fgdc24vi4mzna9icr7-unit-script-ethminer-start[30801]:  m 11:22:04 .ethminer-wrapp 0:00 A0 53.60 Mh { cu0 26.75 | cu1 26.85 }
```

@symphorien thanks for all your help. the program and service are now working with latest code built from source.  ready_for_review

symphorien commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,116 @@
+
+      notify = mkOption {
+        type = types.string;
+        default = "email%40example.org";
```
when there is no sensible default, maybe specify an example instead of a default.
Same for the wallet, I guess.

symphorien commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,116 @@
+
+      stratumPort = mkOption {
+        type = types.string;
+        default = "9999";
```
maybe specify this as an integer instead ?

symphorien commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,116 @@
+
+      wallet = mkOption {
+        type = types.string;
+        default = "0xa7E7030dbFf0Be4695a36C413C299Fd3469E83f3";
```
example instead of default ? what is this address ?

symphorien commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,116 @@
+
+      apiPort = mkOption {
+        type = types.string;
+        default = "-3333";
```
A port starting with a minus ?

nand0p replied:
minus indicated read-only mode for api.  indicated in description.

symphorien commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,116 @@
+
+      recheck = mkOption {
+        type = types.string;
+        default = "1000";
```
maybe make this an integer ?

symphorien commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,116 @@
+      path = [ pkgs.cudatoolkit ];
+      description = "ethminer ethereum mining service";
+      wantedBy = [ "multi-user.target" ];
+
```
does this need to run as root ?

nand0p replied:
added user with specific sudo privs

nand0p replied:
added specific sudoers

symphorien commented:
If the service does not requires special permissions except the nvidia command, then you could:
* remove the user and sudo configuration
* make the service use a DynamicUser
* launch the nvidia command in `ExecStartPost` prefixed with a minus sign (see `man systemd.service`): the command will be executed as root.

This has the advantage of not needing an uid and slightly improving security.

nand0p commented:
@symphorien ready_for_review

symphorien commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,109 @@
+      recheck = mkOption {
+        type = types.int;
+        default = 1000;
+        description = "time to recheck.";
```
please specify the unit (seconds, I imagine ?)

nand0p replied:
milliseconds :)

symphorien commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,109 @@
+
+      serviceConfig = {
+        DynamicUser = true;
+        ExecStartPost = "+${pkgs.linuxPackages.nvidia_x11.bin}/bin/nvidia-smi -pl ${toString cfg.power}";
```
please use `config.boot.kernelPackages` instead of `pkgs.linuxPackages`. This will automatically use a version built with the right kernel.

nand0p replied:
done

symphorien commented on pkgs/tools/misc/ethminer/default.nix:
```diff
@@ -0,0 +1,64 @@
+  cmakeFlags = [
+    "-DHUNTER_ENABLED=OFF"
+    "-DETHASHCUDA=ON"
+    "-DCMAKE_INSTALL_PREFIX=./install"
```
I think that if you remove this line, the installPhase will become unnecessary.

nand0p replied:
sill needs the wrapper with path to nvidia looks like, but was able to eliminate the copy to `$out`

symphorien reviewed (commented):
Thanks. Great work so far. I still have some comments, but they are fairly minor.

nand0p commented on pkgs/tools/misc/ethminer/default.nix:
```diff
@@ -0,0 +1,66 @@
+
+  nativeBuildInputs = [
+    cmake
+    cli11
```
```
CMake Error at ethminer/CMakeLists.txt:17 (find_package):
  Could not find a package configuration file provided by "CLI11" with any of
  the following names:

    CLI11Config.cmake
    cli11-config.cmake

  Add the installation prefix of "CLI11" to CMAKE_PREFIX_PATH or set
  "CLI11_DIR" to a directory containing one of the above files.  If "CLI11"
  provides a separate development package or SDK, be sure it has been
  installed.
```
without it, i get this error

symphorien replied:
Ah my bad. In any case, it should be a buildInput instead of a nativeBuildInput, I think.

nand0p replied:
done

nand0p commented on pkgs/tools/misc/ethminer/binary.nix:
```diff
@@ -0,0 +1,33 @@
+  '';
+
+  dontStrip = true;
+
```
removed

nand0p commented on pkgs/development/tools/misc/cli11/default.nix:
```diff
@@ -0,0 +1,30 @@
+  ];
+
+  cmakeFlags = [
+    "-DCLI11_TESTING=OFF"
```
`doCheck = false` due to not able to have the code find gtest, despite trying several things (comments in code)

symphorien replied:
Hmm indeed this gtest thing is tricky. Here is how I got it right:
```diff
diff --git a/pkgs/development/tools/misc/cli11/default.nix b/pkgs/development/tools/misc/cli11/default.nix
index 87b409562b7..bd735373539 100644
--- a/pkgs/development/tools/misc/cli11/default.nix
+++ b/pkgs/development/tools/misc/cli11/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchFromGitHub, cmake, gtest }:
+{ stdenv, fetchFromGitHub, cmake, python, gtest, boost }:
 
 stdenv.mkDerivation rec {
   name = "cli11-${version}";
@@ -15,23 +15,12 @@ stdenv.mkDerivation rec {
     cmake
   ];
 
-  checkInputs = [
-    gtest
-  ];
-
-
-  # NOTE: disabling tests due to gtest not found issue
-  doCheck = false;
-  cmakeFlags = [
-    "-DCLI11_TESTING=OFF"
-  #  "-DNO_GTEST=ON"
-  #  "-DGTEST_ROOT=${gtest.src}/googletest"
-  #  "-DGTEST_INCLUDE_DIR=${gtest.dev}/include"
-  #  "-DGTEST_HOME=${gtest}"
-  #  "-DCMAKE_CURRENT_BINARY_DIR=${gtest.dev}/bin"
-  #  "-DBUILD_SHARED_LIBS=ON"
-  #  "-DGTEST_LIB_DIR=${gtest.dev}/lib"
-  ];
+  doCheck = true;
+  preConfigure = stdenv.lib.optionalString doCheck ''
+    rm -rf extern/googletest
+    ln -sf ${gtest.src} extern/googletest
+  '';
+  checkInputs = [ boost python ];
 
   enableParallelBuilding = true;
```
The symlink is necessary because of https://github.com/CLIUtils/CLI11/blob/master/tests/CMakeLists.txt#L1

nand0p replied:
fixed

nand0p commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,136 @@
+
+      toolkit = mkOption {
+        type = types.string;
+        default = "cuda";
```
i think it has to be either cuda or opencl, not both. if i am misunderstanding, perhaps an example?

symphorien replied:
```suggestion
        type = types.enum [ "cuda" "opencl" ];
```
(untested)

nand0p replied:
fixed

symphorien reviewed (approved):
Thanks for your PR. Looks good to me.

nand0p commented:
@cpages @ttuegel @Infinisil @symphorien Anything i can do to help get this merged in?  Thanks!

Infinisil commented on pkgs/tools/misc/ethminer/default.nix:
```diff
@@ -0,0 +1,62 @@
+  '';
+
+  postInstall = ''
+    wrapProgram $out/bin/ethminer --set LD_LIBRARY_PATH "${pkgs.linuxPackages.nvidia_x11}/lib"
```
If you need to get it working properly with opengl drivers, you should wrap it with `--prefix LD_LIBRARY_PATH : /run/opengl-driver/lib`. This here would only work for a specific nvidia driver.

Infinisil commented on pkgs/tools/misc/ethminer/default.nix:
```diff
@@ -0,0 +1,62 @@
+{ fetchurl, lib, pkgs, stdenv, fetchFromGitHub, opencl-headers, cmake, jsoncpp, boost, makeWrapper,
```
`pkgs` should never be an argument here, after the change suggested above you can remove it. Also you can remove `lib` and `fetchurl` as these aren't used anywhere.

Infinisil commented on pkgs/development/tools/misc/cli11/default.nix:
```diff
@@ -0,0 +1,35 @@
+
+  doCheck = true;
+
+  preConfigure = stdenv.lib.optionalString doCheck ''
```
This `stdenv.lib.optionalString doCheck` only changes behavior if somebody changes the very line just above this one by setting `doCheck = false`, it won't have any effect when somebody sets `doCheck` from outside. So I think this condition can be removed.

Infinisil commented on pkgs/development/libraries/ethash/default.nix:
```diff
@@ -0,0 +1,59 @@
+
+stdenv.mkDerivation rec {
+  name = "${pkgname}-${version}";
+  pkgname = "ethash";
```
This should be `pname = "ethhash"`.

Infinisil commented on pkgs/tools/misc/ethminer/default.nix:
```diff
@@ -0,0 +1,62 @@
+
+stdenv.mkDerivation rec {
+  name = "${pkgname}-${version}";
+  pkgname = "ethminer";
```
Also `pname` here

Infinisil commented on pkgs/development/libraries/ethash/default.nix:
```diff
@@ -0,0 +1,59 @@
+{ lib, pkgs, stdenv, fetchFromGitHub, cmake, boost, cryptopp, opencl-headers, opencl-info,
```
`lib` and `pkgs` aren't used.

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+      };
+
+      script = ''
+        ${pkgs.ethminer}/bin/.ethminer-wrapped \
```
It's a bit confusing, but `.ethminer-wrapped` is actually the non-wrapped version, `/bin/ethminer` is the wrapped one.

nand0p replied:
Don't we want to have systemd call the unwrapped program directly? (which requires `environment.LD_LIBRARY_PATH="${config.boot.kernelPackages.nvidia_x11}/lib"` in systemd service config)?

Infinisil replied:
Oh I see, makes sense

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+      };
+
+      environment = {
+        LD_LIBRARY_PATH = "${config.boot.kernelPackages.nvidia_x11}/lib";
```
Then you also won't need this.

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+
+      serviceConfig = {
+        DynamicUser = true;
+        ExecStartPost = "+${config.boot.kernelPackages.nvidia_x11.bin}/bin/nvidia-smi -pl ${toString cfg.power}";
```
What does this do and why is it needed? Also this probably only works for nvidia GPU's, so shouldn't this be conditioned on that?

And you should use `getBin config.boot.kernelPackages.nvidia_x11` instead.

nand0p replied:
this throttles the power used by nvidia GPU's, as they do not need the default consumption to achieve full mining ability.  the limit does vary by nvidia model, and the default i set is for the 1070. This is only applicable to nvidia cards, so i will make that conditional.

nand0p replied:
i made this conditional on cuda toolkit being specified `ExecStartPost = optional (cfg.toolkit == "cuda") "+${config.boot.kernelPackages.nvidia_x11.bin}/bin/nvidia-smi -pl ${toString cfg.maxPower}";`

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+      };
+
+      notify = mkOption {
+        type = types.string;
```
All of these types should be `types.str` instead, `types.string` is deprecated.

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+      };
+
+      stratumPort = mkOption {
+        type = types.int;
```
Can use `types.port`

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+
+      wallet = mkOption {
+        type = types.string;
+        example = "0x0484A1EfB3C88B0C7dE21919D480fF51099c8113";
```
Where did you get this example from?

nand0p replied:
https://etherscan.io/address/0x0484A1EfB3C88B0C7dE21919D480fF51099c8113 this is just an example of a real, but unused address.

Infinisil replied:
Eh.. I think it's best to just pick a random number here instead of an actual address

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+      power = mkOption {
+        type = types.int;
+        default = 115;
+        description = "miner max watt usage.";
```
All descriptions should start with a capital letter.

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+        description = "mining rig name.";
+      };
+
+      notify = mkOption {
```
Shouldn't this be called something like `registerMail`?

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+        description = "url encoded email address to register with pool.";
+      };
+
+      power = mkOption {
```
How about `maxPower`

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+        description = "enable ethminer ether mining.";
+      };
+
+      recheck = mkOption {
```
How about `recheckInterval`

infinisil reviewed (changes requested):
You should have a separate commit for each package you add, and a final commit for the NixOS module.

Commit:
ethminer: init at 0.18.0-rc.0

- Tested on NixOS

Commit:
ethash: init at 0.4.2

- Tested on NixOS

Commit:
cli11: init at 1.7.1

- Tested on NixOS

Commit:
jsoncpp: 1.8.4 add cmake support

- Tested on NixOS

Commit:
nixos/ethminer: init

- Tested on NixOS

nand0p commented:
@Infinisil ready_for_review

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+      recheckInterval = mkOption {
+        type = types.int;
+        default = 2000;
+        description = "Time in milliseconds to farm recheck.";
```
I suggest "Interval in milliseconds between farm rechecks." or something like that instead

Infinisil commented:
Oh and the last commit should be named `nixos/ethminer: init`

nand0p commented:
@Infinisil ready_for_review

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+      recheckInterval = mkOption {
+        type = types.int;
+        default = 2000;
+        description = "Interval in milliseconds to farm recheck.";
```
You still haven't changed this.. https://github.com/NixOS/nixpkgs/pull/55422#discussion_r272756608

nand0p replied:
> I suggest "Interval in milliseconds between farm rechecks." or something like that instead

i do think this was changed as recommended?

Infinisil replied:
Ah yeah, sorry you did change it, but I had to mention it again because this sentence sounds wrong to me still, which is why I suggested "between" in the first place.

nand0p replied:
agreed "between" makes it better. fixed.

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+
+      wallet = mkOption {
+        type = types.str;
+        example = "1x0484A1EfB3C88B0C7dE21919D480fF51099c8113";
```
You still haven't changed this! This is getting annoying, I can't trust you like this if you mark everything as resolved when in fact it's still the same. Yes you changed the "0x" to "1x", but "0x" is supposed to signify that it's a hex address, "1x" is invalid for that

nand0p replied:
> a random (invalid) one would be better.
by doing `1x` it becomes invalid.  im not sure how else to ensure an address is invalid?  i will increment the last char, instead of the first, as i think that is what you are suggesting.

Infinisil replied:
Well of course you can't guarantee it, but if you generate a random number, the chance of it corresponding to a valid address is in the order of 0.0000000001% or so. We still want plausible examples though, which an address starting with `1x` is not. `0x01234...` is perfect though, it even signifies that this is an example value, and it's just as improbably to be an actual address as a random number.

nand0p replied:
updated with the more example-like example (`0x0123456789abcdef0123456789abcdef01234567`)

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+          --report-hashrate \
+          --${cfg.toolkit} \
+          --api-port ${toString cfg.apiPort} \
+          --pool stratum1+tcp://${cfg.wallet}@${lib.escapeShellArg cfg.pool}:${toString cfg.stratumPort}/${lib.escapeShellArg cfg.rig}/${lib.escapeShellArg cfg.registerMail}
```
Just use a single `escapeShellArg`: `--pool ${escapeShellArg "stratum1+tcp://${cfg.wallet}@${cfg.pool}:${toString cfg.stratumPort}/${cfg.rig}/${cfg.registerMail}"}`

Even better would be to assign a variable to make this line not as long

nand0p replied:
added `poolUrl` variable for this.

Infinisil commented on nixos/modules/services/misc/ethminer.nix:
```diff
@@ -0,0 +1,114 @@
+
+      serviceConfig = {
+        DynamicUser = true;
+        ExecStartPost = optional (cfg.toolkit == "cuda") "+${config.boot.kernelPackages.nvidia_x11.bin}/bin/nvidia-smi -pl ${toString cfg.maxPower}";
```
You didn't address https://github.com/NixOS/nixpkgs/pull/55422#discussion_r262233463 fully

nand0p replied:
updated to `ExecStartPost = optional (cfg.toolkit == "cuda") "+${getBin config.boot.kernelPackages.nvidia_x11}/bin/nvidia-smi -pl ${toString cfg.maxPower}";`

infinisil reviewed (changes requested):
Don't just mark things as resolved if you haven't resolved them!

ttuegel commented on pkgs/development/libraries/jsoncpp/default.nix:
```diff
@@ -36,13 +33,14 @@ stdenv.mkDerivation rec {
   cmakeFlags = [
     "-DBUILD_SHARED_LIBS=ON"
     "-DBUILD_STATIC_LIBS=OFF"
+    "-DJSONCPP_WITH_CMAKE_PACKAGE=ON"
```
What is this CMake flag for?

nand0p replied:
this provides cmake support for the jsoncpp package, which fixes the following error:

```
CMake Error at CMakeLists.txt:64 (find_package):
  Could not find a package configuration file provided by "jsoncpp" with any
  of the following names:

    jsoncppConfig.cmake
    jsoncpp-config.cmake

  Add the installation prefix of "jsoncpp" to CMAKE_PREFIX_PATH or set
  "jsoncpp_DIR" to a directory containing one of the above files.  If
  "jsoncpp" provides a separate development package or SDK, be sure it has
  been installed.
```

nand0p commented:
@Infinisil ready_for_review
